
<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta charset="utf-8" />
    <title>EpiClass’s documentation! &#8212; EpiClass 2.2.0 documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="_static/language_data.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Documentation" href="Documentation.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="epiclass-s-documentation">
<h1>EpiClass’s documentation!<a class="headerlink" href="#epiclass-s-documentation" title="Permalink to this headline">¶</a></h1>
</div>
<div class="section" id="overview">
<h1>Overview:<a class="headerlink" href="#overview" title="Permalink to this headline">¶</a></h1>
<p>EpiClass</p>
<p>Optimizing and predicting performance of DNA methylation biomarkers using sequence methylation density information.</p>
</div>
<div class="section" id="quick-install">
<h1>Quick Install:<a class="headerlink" href="#quick-install" title="Permalink to this headline">¶</a></h1>
<p>We recommend to make conda environment with python==3.7 first:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">conda</span> <span class="n">create</span> <span class="o">-</span><span class="n">n</span> <span class="n">name</span> <span class="n">python</span><span class="o">==</span><span class="mf">3.7</span> <span class="n">pip</span>

<span class="n">pip</span> <span class="n">install</span> <span class="n">EpiClass</span>
</pre></div>
</div>
</div>
<div class="section" id="installation-from-github">
<h1>Installation from GitHub:<a class="headerlink" href="#installation-from-github" title="Permalink to this headline">¶</a></h1>
<p>A copy of the package can be obtained by downloading the git repository:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">git</span> <span class="n">clone</span> <span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">bmill3r</span><span class="o">/</span><span class="n">EpiClass</span>
</pre></div>
</div>
<p>Required dependencies can be found in epiclass_env.yml.
Importantly, epiclass is meant for python 3.7.</p>
<p>samtools 1.6 was used in its development. It is required for running <cite>epiclass READtoMD</cite> on .bam sequence alignment files.</p>
<p>We first recommend installing in a fresh python virtual environment, either mediated by conda or virtualenv, using the epiclass_env.yml.</p>
<p>For conda:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">cd</span> <span class="n">EpiClass</span>

<span class="n">conda</span> <span class="n">env</span> <span class="n">create</span> <span class="o">-</span><span class="n">f</span> <span class="n">epiclass_env</span><span class="o">.</span><span class="n">yml</span>

<span class="n">conda</span> <span class="n">activate</span> <span class="n">epiclass</span>
</pre></div>
</div>
<p>Then install:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">python</span> <span class="n">setup</span><span class="o">.</span><span class="n">py</span> <span class="n">build</span>

<span class="n">python</span> <span class="n">setup</span><span class="o">.</span><span class="n">py</span> <span class="n">install</span>
</pre></div>
</div>
<p>Alternatively:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">pip</span> <span class="n">install</span> <span class="n">EpiClass</span><span class="o">-*.*.*.</span><span class="n">tar</span><span class="o">.</span><span class="n">gz</span>
</pre></div>
</div>
<p>(found in: /EpiClass/dist/)</p>
<p>Check that epiclass is installed with:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">epiclass</span> <span class="o">-</span><span class="n">V</span>
</pre></div>
</div>
<p>or:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">epiclass</span> <span class="o">-</span><span class="n">h</span>
</pre></div>
</div>
</div>
<div class="section" id="the-vignette">
<h1>The Vignette:<a class="headerlink" href="#the-vignette" title="Permalink to this headline">¶</a></h1>
<p>For a deeper insight into how the code works and generating the manuscript figures, check out the vignette and associated jupyter notebooks:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">bmill3r</span><span class="o">/</span><span class="n">EpiClass</span><span class="o">/</span><span class="n">manuscript_figures</span><span class="o">/</span><span class="n">vignette</span><span class="o">/</span><span class="n">README_vignette</span><span class="o">.</span><span class="n">ipynb</span>
</pre></div>
</div>
</div>
<div class="section" id="quick-usage">
<h1>Quick Usage:<a class="headerlink" href="#quick-usage" title="Permalink to this headline">¶</a></h1>
<div class="section" id="generate-methylation-density-table-dreamtomd-readtomd-dt-timestamp-csv">
<h2>Generate methylation density table ({DREAMtoMD|READtoMD}.DT.{timestamp}.csv)<a class="headerlink" href="#generate-methylation-density-table-dreamtomd-readtomd-dt-timestamp-csv" title="Permalink to this headline">¶</a></h2>
<p>From either sequencing alignment reads or DREAMing methylation melt data.</p>
<div class="section" id="for-sequences">
<h3>For sequences:<a class="headerlink" href="#for-sequences" title="Permalink to this headline">¶</a></h3>
<ol class="lowerroman">
<li><p>Each file must be its own .sam or .bam file. Point to the directory that contains the files to analyze.</p></li>
<li><p>Alignment files must be from Bismark (<a class="reference external" href="https://www.bioinformatics.babraham.ac.uk/projects/bismark/">https://www.bioinformatics.babraham.ac.uk/projects/bismark/</a>).</p></li>
<li><p>Command to process the sequencing reads:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">epiclass</span> <span class="n">READtoMD</span> <span class="o">-</span><span class="n">i</span> <span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">files</span> <span class="o">--</span><span class="n">interval</span> <span class="n">BED</span><span class="o">/</span><span class="k">with</span><span class="o">/</span><span class="n">genome</span><span class="o">/</span><span class="n">coordinates</span>
</pre></div>
</div>
</li>
</ol>
<p>The interval file is a bed file that has genomic coordinates for which methylation density information will be extracted for the samples. If multiple intervals passed, the analysis will be done using pooled information from all of them. Otherwise, can pass “chr:start-stop” for a single genomic region.</p>
</div>
<div class="section" id="for-dreaming-data">
<h3>For DREAMing data:<a class="headerlink" href="#for-dreaming-data" title="Permalink to this headline">¶</a></h3>
<ol class="lowerroman">
<li><p>Use a raw melt temp .csv file as input. A custom template is provided.</p></li>
<li><p>Command to process the DREAMing melt data:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">epiclass</span> <span class="n">DREAMtoMD</span> <span class="o">-</span><span class="n">i</span> <span class="n">rawmelt</span><span class="o">.</span><span class="n">csv</span> <span class="o">-</span><span class="n">temps</span> <span class="n">tempsToMDs</span><span class="o">.</span><span class="n">csv</span> <span class="o">-</span><span class="n">cpg</span> <span class="n">numberCpGs</span>
</pre></div>
</div>
</li>
</ol>
<p>tempsToMDs.csv should be a two column file indicating the methylation density a melting temperature corresponds to for the given locus for which DREAMing was performed. numberCpGs indicates the number of internal CpGs within the given locus for which DREAMing was performed. Both of these are required.</p>
</div>
</div>
<div class="section" id="methylation-density-binary-classifier-mdbc-performance">
<h2>Methylation Density Binary Classifier (MDBC) performance<a class="headerlink" href="#methylation-density-binary-classifier-mdbc-performance" title="Permalink to this headline">¶</a></h2>
<p>Compute the optimal methylation density cutoff for the case and control samples, given the sequence methylation density profiles for the genomic region of interest (or pooled genomic regions:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">epiclass</span> <span class="n">MDBC</span> <span class="o">-</span><span class="n">i</span> <span class="n">densityTable</span><span class="o">.</span><span class="n">csv</span> <span class="o">-</span><span class="n">a</span> <span class="n">cases</span> <span class="o">-</span><span class="n">b</span> <span class="n">controls</span>
</pre></div>
</div>
<p>By default, returns summary tables containing the TPR, 1 - FPR, AUC, and optimal read cutoff for each methylation density cutoff (MDC). Will also return ROCs and boxplots for the optimal MD.</p>
<p>This is done using either the sample read counts or the sample read fractions. As in, the number of reads with a given methylation density or higher, or the sample fraction of reads with a given methylation density or higher.</p>
<p>More extensive plots and sample information can be obtained with additional command flags.</p>
<p>Samples can be normalized based on relative sample fractions, and specific methylation density cutoffs can also be used.</p>
</div>
<div class="section" id="for-more-information">
<h2>For more information:<a class="headerlink" href="#for-more-information" title="Permalink to this headline">¶</a></h2>
<p>use:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">epiclass</span> <span class="o">-</span><span class="n">h</span>

<span class="n">epiclass</span> <span class="n">READtoMD</span> <span class="o">-</span><span class="n">h</span>

<span class="n">epiclass</span> <span class="n">DREAMtoMD</span> <span class="o">-</span><span class="n">h</span>

<span class="n">epiclass</span> <span class="n">MDBC</span> <span class="o">-</span><span class="n">h</span>
</pre></div>
</div>
<div class="section" id="troubleshooting">
<h3>Troubleshooting:<a class="headerlink" href="#troubleshooting" title="Permalink to this headline">¶</a></h3>
<p>Overall:
1. Using full, absolute paths is probably a good idea…</p>
</div>
<div class="section" id="even-more-detailed-information">
<h3>Even more detailed information:<a class="headerlink" href="#even-more-detailed-information" title="Permalink to this headline">¶</a></h3>
<p>Follow the <cite>Documentation</cite> link under ‘Contents” below:</p>
<div class="toctree-wrapper compound">
<p class="caption"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="Documentation.html">Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="License.html">License</a></li>
<li class="toctree-l1"><a class="reference internal" href="Need Help.html">Need Help</a></li>
</ul>
</div>
</div>
</div>
</div>
<div class="section" id="indices-and-tables">
<h1>Indices and tables<a class="headerlink" href="#indices-and-tables" title="Permalink to this headline">¶</a></h1>
<ul class="simple">
<li><p><a class="reference internal" href="search.html"><span class="std std-ref">Search Page</span></a></p></li>
</ul>
</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="#">EpiClass</a></h1>








<h3>Navigation</h3>
<p class="caption"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="Documentation.html">Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="License.html">License</a></li>
<li class="toctree-l1"><a class="reference internal" href="Need Help.html">Need Help</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="#">Documentation overview</a><ul>
      <li>Next: <a href="Documentation.html" title="next chapter">Documentation</a></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2019, Brendan F. Miller.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 2.2.0</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/index.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>